src/Workspaces/Core/Portable/Workspace/Solution/SolutionCompilationState.CompilationTracker.cs

```diff
@@ -783,6 +637,78 @@ private Compilation CreateEmptyCompilation()
                 return finalState.HasSuccessfullyLoaded;
             }
 
+            public ICompilationTracker FreezePartialState(CancellationToken cancellationToken)
+            {
+                var state = this.ReadState();
+
+                // If we're finalized and already frozen, we can just use ourselves.
+                if (state is FinalCompilationTrackerState { IsFrozen: true } finalState)
+                    return this;
+
+                var projectState = state switch
+                {
+                    // If we don't have an existing state, then transition to a project state without any data.
+                    null => this.ProjectState.RemoveAllDocuments(),
+                    FinalCompilationTrackerState => this.ProjectState,
+                    // If we have an in progress state with no steps, then we're just at the current project state.
+                    InProgressState { PendingTranslationSteps: [] } => this.ProjectState,
+                    // Otherwise, reset us to whatever state the InProgressState had currently transitioned to.
+                    InProgressState inProgressState => inProgressState.PendingTranslationSteps.First().oldState,
```

Comment:
@jasonmalinowski  i def want to understand this list, and the `oldState` in it.  I keep on partially convincing myself that it should not be needed (which would simplify a bunch of code as well).  i'd love it if an InProgressState was nothing more than a list of compilation/driver transformations.

Reply:
The short answer I think is it means when you want to grab a frozen partial, and we grab the current compilation we have, you already have a ProjectState that matches the full document contents of it. You could imagine the transformations also containing the code to change ProjectStates too, but then you're duplicating that logic that already exists in the main code for the Add/Remove/Whatever methods.